Combine license matches in new LicenseDetection

AyanSinhaMahapatra opened the pull request:
See #2878 

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

AyanSinhaMahapatra committed:
Import functions from `scancode-analyzer`

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify LicenseDetection attributes

Remove primary license expression attributes, move matches to last.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Create LicenseDetection serialization function

Move the LicenseMatch serialization functions into
the LicenseDetection function.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Enable LicenseDetection in API

Create LicenseDetection functions and enable them in the API.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove the --unknown-licenses option

Removes the unknown licenses CLI option, as this should be
default behaviour.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update local license dereferencing code

Update local unknown license reference code to use
LicenseDetection functions.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update unknown license dereferencing for key-files

This updates the unknown license dereferencing code to also
look for the referenced filename in root of the codebase.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update intro rules with is_license_intro as True

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add test files for LicenseDetection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add LicenseDetection full scan tests

Adds tests for:

- unknown license intros
- unknown local license references

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add function to create index from test rules folders

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on src/licensedcode/cache.py:
```diff
@@ -138,6 +138,40 @@ def load_or_build(
             raise
 
 
+class MockCacheFromTestRules(LicenseCache):
```
Should this be code in a tests utility rather than here?

AyanSinhaMahapatra replied:
Yes, moving this.

pombredanne commented on src/licensedcode/cache.py:
```diff
@@ -353,6 +387,14 @@ def get_index(force=False, index_all_languages=False):
     return get_cache(force=force, index_all_languages=index_all_languages).index
 
 
+def get_index_with_test_rules(test_rules_dir):
```
Same comment as above: why not moving this to tests?

AyanSinhaMahapatra replied:
Yes, moving this.

pombredanne commented on src/licensedcode/data/rules/lead-in_unknown_2.yml:
```diff
@@ -1,3 +1,3 @@
 license_expression: unknown-license-reference
-is_license_notice: yes
+is_license_intro: yes
```
Good catch!

pombredanne commented on src/scancode/api.py:
```diff
@@ -129,18 +129,11 @@ def get_urls(location, threshold=50, **kwargs):
     return dict(urls=results)
 
 
-SPDX_LICENSE_URL = 'https://spdx.org/licenses/{}'
```
If you remove these from the api module. you need to document this in the CHANGELOG or provide some way to get this from API?

AyanSinhaMahapatra replied:
Moving these back into the API and importing from there.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -24,34 +27,97 @@
+DEJACODE_LICENSE_URL = 'https://enterprise.dejacode.com/urn/urn:dje:license:{}'
+SCANCODE_LICENSEDB_URL = 'https://scancode-licensedb.aboutcode.org/{}'
+
+# All values of match_coverage less than this value are taken as
```
less of higher?

AyanSinhaMahapatra replied:
Adding more explanation docs.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -24,34 +27,97 @@
+
+# How many Lines in between has to be present for two matches being of a different group
+# (i.e. and therefore, different rule)
+LINES_THRESHOLD = 4
```
Should this be synced with https://github.com/nexB/scancode-toolkit/blob/19d77e99c0c37270752849a1f27bb3ec9ad96c72/src/licensedcode/query.py#L196 and should be a single constant for sanity?

AyanSinhaMahapatra replied:
Yes, creating the constant there and importing instead.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -24,34 +27,97 @@
+
+# Threshold Values of start line and rule length for a match to likely be a false positive
+# (more than the start_line threshold and less than the rule_length threshold)
+FALSE_POSITIVE_START_LINE_THRESHOLD = 1000
```
This may require a bit more explanation

AyanSinhaMahapatra replied:
Added.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -24,34 +27,97 @@
+USE_LICENSE_CASE_BERT_MODEL = False
+USE_FALSE_POSITIVE_BERT_MODEL = False
+
+ISSUE_CASES_VERSION = 0.1
```
What is this?

AyanSinhaMahapatra replied:
This should not have been moved in from the Analyzer, removing.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -24,34 +27,97 @@
+FALSE_POSITIVE_RULE_LENGTH_THRESHOLD = 3
+
+# Whether to Use the NLP BERT Models
+USE_LICENSE_CASE_BERT_MODEL = False
```
Could these be instead used if present and contributed by an optional plugin?

AyanSinhaMahapatra replied:
Removing for now, this shouldn't have been brought in.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -63,44 +129,13 @@ class LicenseDetection:
-            'using the SPDX license expression syntax and SPDX license ids.')
-    )
-
     spdx_license_expression = attr.ib(
```
This likely should be a computed field.

AyanSinhaMahapatra replied:
This is computed from the `license_expression`, but as this is also returned as an attribute in the LicenseDetection level we need to keep it here?

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
     )
 
+    matches = attr.ib(
+        #repr=matches_compact_repr,
```
```suggestion
```

AyanSinhaMahapatra replied:
Removed.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+    def from_matches(cls, matches):
+        """
+        Return a LicenseDetection created out of `matches` list of LicenseMatch
+        objects after computing the license_expression by analyzing and combining
```
```suggestion
```

AyanSinhaMahapatra replied:
Modified.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+        """
+        Return a LicenseDetection created out of `matches` list of LicenseMatch
+        objects after computing the license_expression by analyzing and combining
+        matches for the different cases.
```
```suggestion
```

AyanSinhaMahapatra replied:
Yes.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+    @classmethod
+    def from_matches(cls, matches):
+        """
+        Return a LicenseDetection created out of `matches` list of LicenseMatch
```
Less is more here: let's not expose too much of the sausage making in the docstring.
```suggestion
        Return a LicenseDetection created out of `matches` list of LicenseMatch.
```

AyanSinhaMahapatra replied:
Added.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+        """
+        reasons, license_expression = get_detected_license_expression(
+            matches=matches,
+            analysis=analyze_detection(matches),
```
What if the analysis is not conducive to creating a proper detection?

AyanSinhaMahapatra replied:
The license-expression is `None` and matches for this part is added to the `license_clues` instead of `license_detections`

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+            analysis=analyze_detection(matches),
+        )
+
+        if license_expression == None:
```
```suggestion
        if license_expression is None:
```

AyanSinhaMahapatra replied:
Added.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+        )
+
+        if license_expression == None:
+            return cls(matches=matches)
```
So we end up having a detection that contains no expression, and is therefore inconclusive?

AyanSinhaMahapatra replied:
Yes, this is added to `license_clues` as discussed above.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+    def query(self):
+        # A LicenseDetection will always be created with matches
+        if self.matches:
+            # All the matches in a file or in a LicenseDetection point to the
```
Are you sure this is guaranteed? I would feel more comfy with an `assert` to protect against this and validate we have indeed only one.

AyanSinhaMahapatra replied:
Yes, putting an assert.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+    @property
+    def query(self):
+        # A LicenseDetection will always be created with matches
+        if self.matches:
```
Can a detection ever be create from empty matches?

AyanSinhaMahapatra replied:
No, this will never happen.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
 
+    matches = attr.ib(
+        #repr=matches_compact_repr,
+        default=attr.Factory(list),
```
You could add a validator to ensure this is not empty

AyanSinhaMahapatra replied:
This won't happen, so I added another check on creation?

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+        return cls(
+            matches=matches,
+            license_expression=str(license_expression),
+            spdx_license_expression=str(spdx_license_expression),
```
IMHO we should remove `spdx_license_expression` and make this a computed  property
This will make the code simpler all around.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+    
+    @property
+    def qspans(self):
+        return [match.qspan for match in self.matches]
```
Would not you rather have a single merged span here?

AyanSinhaMahapatra replied:
Actually we would then again merge across all detections and then compute the percentage license text. Isn't it better do this once?

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+    @property
+    def identifier_with_text(self):
+        """
+        This is an identifier for a issue, which is an unknown license intro,
```
`unknown license intro` ? what is this doing here? I am not sure I get why you mention this

AyanSinhaMahapatra replied:
Should not be mentioned. Basically in cases of small unknown matches, the matched text is required to create a unique identifier, otherwise the rule identifier and the coverage is enough in most cases. But as we discussed we need something absolute so this will change anyway.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+        return [match.qspan for match in self.matches]
+
+    @property
+    def identifier(self):
```
Where are these identifiers used and what is your intention? Is this just to create something unique? or to create some to check that two detections are the same?
If the later, then you should rather have a function that checks equivalence explicitly rather than having indirection with an identifier.

AyanSinhaMahapatra replied:
> create something unique

This yes. As the code for finding unique part doesn't check between two detections, rather gets unique ones from a list of identifiers.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -118,12 +153,90 @@ class LicenseDetection:
+        """
+        data = []
+        for license_match in self.original_licenses:
+            identifier = (license_match.rule_identifier, license_match.coverage(),)
```
IMHO coverage is not a good elements of identification. Instead the rule_identifier + matched ispan would be best if you are looking for some kind of content-based id... But this would still be missing that some extra unknown or stop words may exist on the qspan-side within and inside the range of ispan... so this needs a bit more detailing.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -155,7 +268,7 @@ def relevance(self):
         rule length.
         """
-        return compute_relevance(self.rules_len)
+        return compute_relevance(self.rules_length)
```
Are you sure? `self.rules_length` is a  function 
```suggestion
        return compute_relevance(self.rules_length())
```

AyanSinhaMahapatra replied:
Yes, thanks. Updating this.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+):
+    """
+    Return a list of "licenses" scan data built from a license match.
+    Used directly only internally for testing.
```
May be this is best in the tests/  then?

AyanSinhaMahapatra replied:
This was being used in the serialization actually earlier too, still being used. Will update  the docstring.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+    return detected_licenses
+
+
+def matches_compact_repr(matches):
```
This seems not used anywhere

AyanSinhaMahapatra replied:
Yes this was present earlier in the code. Removing this.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+
+def is_correct_detection(license_matches):
+    """
+    Return True if all the license matches in a file-region are correct
```
Do not expose too much of the internal implementation in your docstring,
```suggestion
    Return True if all the matches in ` license_matches` are correct e.g., highly accurate and not ambiguous.
```

AyanSinhaMahapatra replied:
Yes.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+def is_correct_detection(license_matches):
+    """
+    Return True if all the license matches in a file-region are correct
+    license detections, as they are either SPDX license tags, or the file content has
```
```suggestion
    license detections, as they are either SPDX license tags, or the file content has
```

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+    """
+    Return True if all the license matches in a file-region are correct
+    license detections, as they are either SPDX license tags, or the file content has
+    a exact match with a license hash.
```
```suggestion
    a exact match with a license hash.
```

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+    coverage_values = (
+        license_match.coverage() for license_match in license_matches
+    )
+    return any(coverage_value < threshold for coverage_value in coverage_values)
```
IMHO you can conflate the two lines without loosing redability
```suggestion
    return any(license_match.coverage() < threshold for for license_match in license_matches)
```

AyanSinhaMahapatra replied:
Updated. Thanks!

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+    Calculates a `query_coverage_coefficient` value for that match. For a match:
+    1. If this value is 0, i.e. `score`==`match_coverage`*`rule_Relevance`, then
+       there are no extra words in that license match.
+    2. If this value is a +ve number, i.e. `score`!=`match_coverage`*`rule_Relevance`,
```
`+ve`?

pombredanne replied:
Also can you give an example?

AyanSinhaMahapatra replied:
Updated and see https://github.com/nexB/scancode-analyzer/blob/main/tests/test_license_analyzer.py#L124 and https://github.com/nexB/scancode-analyzer/blob/main/tests/data/analyzer/analyzer_is_extra_words_true_one.json for an example.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+
+def is_extra_words(license_matches):
+    """
+    Return True if any of the license matches in a file-region has extra words. Having
```
Let's not talk about file region since this is not a concept here in this function
```suggestion
    Return True if any of the license matches has extra words. Having
```

AyanSinhaMahapatra replied:
Modifying.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+    return score_coverage_relevance - license_match.score()
+
+
+def is_extra_words(license_matches):
```
```suggestion
def has_extra_words(license_matches):
```

AyanSinhaMahapatra replied:
Modified.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+    """
+    Return True if all the license matches is an unknown intro LicenseMatch.
+
+    :param license_matches: list
```
I see you started using this style of docstring... here this is incorrect: this is a not  list . Also this feels superfluous in this simple case and a few other cases where the meaning of arguments is obvious.

AyanSinhaMahapatra replied:
Yes, removing this style and updating the docstrings to reflect the updates.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -263,6 +370,278 @@ def matched_text(
+
+
+def is_license_clues(license_matches):
+    """
```
Is this instead a method of a LicenseDetection?

AyanSinhaMahapatra replied:
Same as https://github.com/nexB/scancode-toolkit/pull/2961#discussion_r877480348

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+    non-empty `referenced_filename`, i.e. contains a license
+    reference to a local file.
+    """
+    if license_match['matched_rule']['referenced_filenames']:
```
```suggestion
    return bool(license_match['matched_rule']['referenced_filenames'])
```

AyanSinhaMahapatra replied:
Yes.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+    a local file like "see license in LICENSE.md". In these cases we can
+    remove the unknown reference with the license detection in the file.
+    """
+    return [match for match in license_matches if not is_license_reference_local_file(match)]
```
IMHO inlining is good enough for a single use
```suggestion
    return [match for match in license_matches if not license_match['matched_rule']['referenced_filenames']]
```

AyanSinhaMahapatra replied:
We are having all these functions in case they are reused?

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+    return False
+
+
+def combine_license_references(license_matches):
```
This is not combining, but only filtering

AyanSinhaMahapatra replied:
Changing this to filtering.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+    references to local files with licenses.
+
+    A common source of unknown license detections in license references to
+    a local file like "see license in LICENSE.md". In these cases we can
```
But what is "these cases" here? We are saying too much or too litlle. IMHO too much.

AyanSinhaMahapatra replied:
Removing.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+        return "correct-license-detection"
+
+    elif has_unknown_intro_before_detection(license_matches):
+        return "unknown-intro-before-detection"
```
You need to create constants for these

AyanSinhaMahapatra replied:
Creating Enum like the combination reasons.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+
+def get_detected_license_expression(matches, analysis):
+    """
+    Get `reasons` and `combined_license_expression` after combining
```
```suggestion
    Return a tuple of (reasons, combined_expression) by combining a `matches` list of LicenseMatch objects using an `analysis` code string.
```

AyanSinhaMahapatra replied:
Adding.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+    matches_for_expression = None
+    reasons = []
+
+    if analysis == "unknown-intro-before-detection":
```
Use constants

AyanSinhaMahapatra replied:
Yes, modified.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+
+    if isinstance(matches[0], dict):
+        combined_expression = combine_expressions(
+            expressions=[match['matched_rule']['license_expression'] for match in matches_for_expression]
```
Should you rather use deserialized real LicenseMatch objects throughout and rules recreated from the scan data?
This would make the whole code cleaner rather than using dicts and key lookups.

AyanSinhaMahapatra replied:
This would make the code at process_codebase much more complicated tho, as we use have to serialize and de-serialize all the detections (and matches in them?)

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
+    """
+    license_matches = []
+    for detection in license_detections:
+        license_matches.extend(detection["matches"])
```
matches should be an attribute and we should not carry around a bunch of dicts IMHO

pombredanne reviewed (changes requested):
Thanks! that's quite a piece of work! :bow:  I provided some comments, but I did not review everything just yet. Of note you should use LicenseMatch and objects everywhere and not carry around dictionaries of scan data.
Instead convert these at the boundaries as early as possible deserialized to objects.

pombredanne commented on tests/licensedcode/data/plugin_license/license-expression/scan.expected.json:
```diff
@@ -1,78 +1,62 @@
     {
       "path": "apache-1.0.txt",
       "type": "file",
       "licenses": [
```
IMHO rename this to `license_detections` for sanity and avoid too much user confusion.

AyanSinhaMahapatra replied:
This makes sense. Will update this.

pombredanne commented on tests/licensedcode/data/plugin_license/license-expression/scan.expected.json:
```diff
@@ -1,78 +1,62 @@
-            "rule_relevance": 100
-          }
+          "license_expression": "apache-1.0",
+          "spdx_license_expression": "None",
```
Never None for sure.

AyanSinhaMahapatra replied:
This has been fixed.

pombredanne commented on tests/licensedcode/data/plugin_license/license-expression/scan.expected.json:
```diff
@@ -1,78 +1,62 @@
-          }
+          "license_expression": "apache-1.0",
+          "spdx_license_expression": "None",
+          "combination_reasons": [],
```
I have no idea what this could mean. Try to find a better more obvious name for this attribute. Also why is this empty?

AyanSinhaMahapatra replied:
Should this be populated even if we don't do any combination in the LicenseDetection level?

Not keeping this empty would mean adding `not-combined` at all the `combination_reasons` which didn't have one. 

Yes, this could be named better.

pombredanne commented on tests/licensedcode/data/plugin_license/license-expression/scan.expected.json:
```diff
@@ -1,78 +1,62 @@
+          "combination_reasons": [],
+          "matches": [
+            {
+              "key": "apache-1.0",
```
A match is surely not for a key but for an expression (that's a problem in the current licenses output) also we do not want to repeat forever details of licenses which should be pushed out in the --license-reference . Also since a match is for a single rule, this would be best to streamline this rather than having a nested matched_rule. Along these lines, reporting matched_rule reference details only once could help too may be with a new --license-rule-reference that would report the details once

AyanSinhaMahapatra replied:
Yes, this is now pushed, though it is WIP.

pombredanne commented on tests/licensedcode/data/plugin_license/license-expression/scan.expected.json:
```diff
@@ -1,78 +1,62 @@
         }
       ],
+      "license_clues": [],
       "license_expressions": [
```
We cannot double dip: if you have detections, this must go.

AyanSinhaMahapatra replied:
Not sure what this means, `license_detections` and `license_clues` are two seperate lists that are added btw. One is a list of LicenseDetection and another is a list of LicenseMatch for which we couldn't create detections.

pombredanne commented on tests/licensedcode/data/plugin_license/license-expression/scan.expected.json:
```diff
@@ -1,78 +1,62 @@
       "license_expressions": [
         "apache-1.0"
       ],
+      "spdx_license_expressions": [
```
Same as before: this likely needs to go away,  and never  "None"

AyanSinhaMahapatra replied:
This is never None, that part is fixed. 

Do you mean we should only have `spdx_license_expressions` at resource level and not have individual `spdx_license_expression` for each detection? That could make sense yes.

pombredanne reviewed (changes requested):
Here is some feedback on the format. We need to streamline this to the essence and avoid repeating details to remove noise when possible.

AyanSinhaMahapatra committed:
Remove unwanted code, update docstrings

Updates implementation to address review comments.

- Remove code that wasn't being used
- Update docstrings
- Use constants and Enums
- Fix case where local reference can't be found
- Move back constants to API
- Modify test expectations
- Other misc changes

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Move LicenseDetection tests to new file

Also adds a test without the --license-references option.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify LicenseMatch data in results #2416

LicenseMatch data now is based on a license-expression instead
of a license key.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add --licenses-reference option

Add new command line option to not inline license, licenseDB and license
detection level information.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add docs for LicenseDetection and detail referencing

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Do not summurize license detections

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Apply LicenseDetection everywhere

Modify code to align to the LicenseDetection model everywhere.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regenerate test expectations for LicenseDetection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into add-license-detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix SPDX and other output plugins

Modify and update spdx and other output plugins to use LicenseDetection
output data format for licenses.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Rename license fields for resource

* Rename `licenses` to `license_detections`
* Rename `license_expressions` to `detected_license_expression`
* Rename `spdx_license_expressions` to `detected_license_expression_spdx`

Also fix other uses of these attributes around scancode.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expectations for license field renaming

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Rename and Add package license attributes

* Rename `declared_license` to `extracted_license_statement`
* Add `license_detections`
* Rename `license_expressions` to `declared_license_expression`
* Add `detected_license_expression_spdx`

Also add related functions to populate these fields after
package parsing.

Reference: https://github.com/nexB/scancode-toolkit/issues/2065
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add functions for package LicenseDetection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify package parsers to adopt new LicenseDetection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Align to package LicenseDetection

MOdify code to align to the new attribute names for
package LicenseDetection.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify system packages to use LicenseDetection

Modify licence detection code for system packages and build
scripts to support LicenseDetection.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add new license key `undetected-license`

Adds a new license key undetected-license as a probable substitute
for `unknown-license` for package license detection cases.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regenerate test expectations for package LicenseDetection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into add-license-detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne @JonoYang This is ready to review!

JonoYang commented:
@AyanSinhaMahapatra I'll need a little more time to review this, but I think it's looking good so far. You have some failing CI tests, maybe some other test expectations need to be updated?

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -400,60 +400,67 @@ def licenses_data_from_match(
     SCANCODE_LICENSE_DATA_URL = SCANCODE_BASE_URL + '/{}.yml'
 
-    detected_licenses = []
+    result = {}
```
The docstring is now incorrect.
Also I wonder if using an object rather than a dict would be cleaner?

AyanSinhaMahapatra replied:
Updated the docstring. As we'll be converting to mappings anyway and there are a combination of calling functions and attributes and some other processing I'm not sure if we should use an object, what do you suggest here? Looking into this more.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -95,6 +96,18 @@ class CombinationReason(Enum):
+    containing some license information (separated by code/text/others in between),
+    and identified by a start line and an end line.
+    """
+    path = attr.ib(type=str)
```
Is this a relative path or an absolute location?

AyanSinhaMahapatra replied:
This is the same as `resource.path` so it depends on CLI options full/strip root. Not the absolute location, as that is `resource.location`.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -162,6 +183,19 @@ def from_matches(cls, matches):
         )
+    
+    @classmethod
+    def from_mapping(cls, detection):
```
How is this different from calling `LicenseDetection(**detection)` ?

AyanSinhaMahapatra replied:
Yes, removed this.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -181,36 +215,65 @@ def query(self):
         license matches.
+
+        This is not guaranteed to be unique for a detection, as certain small and
+        unknown matches could have the same value for this identifier, but will
```
You should also use the ispan Span on the matched rule side. Together with rule_identifier this is guaranteed to uniquely identify a match in abstraction of where it was matched. And the coverage may not be as useful.... rather the score would be better as this would take into account unknown words that may exist in the matched range. So I would rather use something such as: `((license_match.rule_identifier, license_match.ispan, license_match.score(),)`

AyanSinhaMahapatra replied:
Here, we don't have the objects as this is after the `process_codebase` step and we have only mappings. We have to then recreate objects here from the mappings. Will go ahead with this.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -373,6 +436,7 @@ def to_dict(
 
         detection = attr.asdict(self)
         detection["matches"] = data_matches
+        _file_region = detection.pop('file_region')
```
attr.asdict can filter attributes directly.... IMHO no need to pop. 

Also:
- why the line 424 above with ` matches = self.matches` ... I am not sure this is needed.
- could the function ` licenses_data_from_match(` better be a LicenseMatch method?

AyanSinhaMahapatra replied:
Yes, making this LicenseMatch method instead. Also using filters now, thanks this is much better.

pombredanne commented on src/licensedcode/plugin_licenses_reference.py:
```diff
@@ -55,30 +63,187 @@ def is_enabled(self, licenses_reference, **kwargs):
+
+            # Get license_expressions from both package and license detections
+            license_licexps = getattr(resource, 'license_expressions', []) or []
+            package_data = getattr(resource, 'package_data', []) or []
```
You are missing out the top level packages IMHO

AyanSinhaMahapatra replied:
We do this too now, and take a set out of the list of license_expression so we only have the unique ones.

pombredanne commented on src/licensedcode/plugin_licenses_reference.py:
```diff
@@ -55,30 +63,187 @@ def is_enabled(self, licenses_reference, **kwargs):
-
-        license_keys = set()
+        """
+        Get Licenses and LicenseDB data from all license detections in a codebase level list
```
I am confused about this method and it feels like you may be trying to do too much at once? Also you may be collecting data multiple time, once for each duplicate?

AyanSinhaMahapatra replied:
We have to go at every occurrence and remove those as we are de referencing these, also we are taking a set at last so we don't have any duplicates. As we are taking expressions from multiple places, we are performing the set operation at last, and there won't be any duplicates at last.

pombredanne commented on src/licensedcode/plugin_licenses_reference.py:
```diff
@@ -55,30 +63,187 @@ def is_enabled(self, licenses_reference, **kwargs):
+
+def get_licensedb_references(license_db_data):
+    """
+    """
```
Either put a docstring or not, but not an empty one

AyanSinhaMahapatra replied:
Added this and others.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -181,36 +215,65 @@ def query(self):
-        for license_match in self.original_licenses:
-            identifier = (license_match.rule_identifier, license_match.coverage(),)
+        for match in self.matches:
+            identifier = (match['licensedb_identifier'], match['match_coverage'],)
```
What is this new licensedb_identifier?

AyanSinhaMahapatra replied:
This has been reverted back to the original `rule_identifier` as discussed.

pombredanne reviewed (changes requested):
Thanks! Here is a first batch of review comments.

pombredanne reviewed (commented):
Note I reviewed commit-by-commit starting with the oldest and that's unlikely the better approach. I am restarting reviewing looking at the final results instead

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -571,7 +571,7 @@ def validate(licenses, verbose=False, no_dupe_urls=False):
 
             if lic.is_unknown:
-                if not 'unknown' in lic.key:
+                if not 'unknown' in lic.key and not 'undetected' in lic.key:
```
IMHO we should introduce soon after this a is_something flag for these licenses rather than to special case their name.

AyanSinhaMahapatra replied:
We didn't add this after all, so resolving this.

pombredanne commented on src/licensedcode/plugin_license.py:
```diff
@@ -11,37 +11,49 @@
+from licensedcode.detection import get_detected_license_expression
+from licensedcode.detection import get_matches_from_detections
+from licensedcode.detection import DetectionCategory
 from commoncode.resource import clean_path
```
This likely would benefit from sorting

AyanSinhaMahapatra replied:
Done.

pombredanne commented on src/licensedcode/plugin_licenses_reference.py:
```diff
@@ -39,7 +43,10 @@ class LicensesReference(PostScanPlugin):
-    codebase_attributes = dict(licenses_reference=attr.ib(default=attr.Factory(list)))
+    codebase_attributes = dict(
+        license_references=attr.ib(default=attr.Factory(list)),
+        licensedb_references=attr.ib(default=attr.Factory(list))
```
I would like to discuss this licensedb_references further ... I am not sure I get it entirely.

AyanSinhaMahapatra replied:
Here there are two levels of de-referencing. One is licenses with a unique license-key, the another is a matched rule/license with a unique rule identifier.

pombredanne commented on src/packagedcode/models.py:
```diff
@@ -1130,6 +1136,33 @@ def create_default_package_data(cls, **kwargs):
+        object, and add the declared_license_expression (and the spdx expression)
+        and corresponding LicenseDetection data.
+        """
+        if not package_data.declared_license_expression and package_data.extracted_license_statement:
```
What about inverting and splitting the ifs for readability?
```suggestion
        if package_data.declared_license_expression:
            return
        if not package_data.extracted_license_statement:
            return
```

AyanSinhaMahapatra replied:
Yes, added this.

pombredanne reviewed (changes requested):
Here is another comment. I think we are also missing a CHANGELOG entry (for the next release)

AyanSinhaMahapatra committed:
Fix pypi setup.py email bug

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add manifest license references detection

Add feature to follow license references present in extracted license
statements from package manifests and add license detections from the
referenced files.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add license from file if empty manifest license

Add feature to add license expression and detections from license
detected at file, if there is no extracted license-statement
and detected license at package_data.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add feature to get package license from sibling file

Adds license detection and expressions from a sibling License/Readme
file if there is no license detected from the package manifest itself.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Allow package LicenseDetection without --licenses

Adds functions to add license detections and expression to packages
without requiring to run scans with --licenses CLI option.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regen datadriven LicenseDetections

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Revert to `rule_identifier`

Rename from `licensedb_identifier` to the previously used attribute
`rule_identifier` as this was wrong.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove `undetected-license` in favour of `unknown`

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regen test expectations and fix tests

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Reorder license expressions and detections in result

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add fucntions to test license detection with subset of rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix datadriven license test errors

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix debian_copyright test failure

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Address review feedback

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add tests from eclipse foundation issues

Fix a bug in LicenseDetection analysis and add tests from issues
reported by the eclipse foundation.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into add-license-detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Regenerate test expectations after merging develop

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add `other_license*` attributes for packages #2065

- Add `other_license_expression` attribute
- Add `other_license_expression_spdx` attribute
- Add `other_license_detections` attribute

For debian copyright files `declared_license_expression` is
set to the primary license, if present, and other licenses
are set to the `other_license_expression` introduced here.

Reference: https://github.com/nexB/scancode-toolkit/issues/2065
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove `compute_normalized_license` functions

Removes previously used compute_normlized_license functions
in favour of the newer populate_license_fields functions.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add `default_license_relation` attribute to handlers

Adds a new attribute deafult_license_relation to datafile
handlers to customize how declared_license_expression is
computed from multiple license detections.

Also fixes phpcomposer license detection by this feature.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Support NuGet license URLs #3037

* Extends the SPDX matcher to also support NuGet license URLs
  these are essentially SPDX license expressions with a different
  introductory text.

* Adds tests for NuGet license URL support

* Deletes (~2K) rules which supported NuGet license URL detection

Reference: https://github.com/nexB/scancode-toolkit/issues/3037
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix rpm tests

Fix a bug in unknown LicenseDetection generation where the query_string
consisted of unknown tokens and this caused failures in matched_text
generation. Fixed by prefixing the query_string with `license`.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into add-license-detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures and expectations after merging develop

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Do not return empty dict as exctracted license

Fix an issue where empty dictionary was being returned as
extraced_license_statement in pypi.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix csv output after adding LicenseDetection`

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne All green!

AyanSinhaMahapatra committed:
Tag intro rule properly

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix package license expression None bug

* Fixes a bug where package license expression was None
* Fixes package license expression derivation by using the LicenseDetection
  objects properly

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Also classify license intro in false positives list

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into add-license-detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Make LicenseMatch grouping affected by presence of license intro

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify license references to also effect clues

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Restore returning whole_lines by default

This restores returning whole_lines behaviour and also adds
attributes to functions so it can be turned off in the case of
debian copyrights where this was problematic.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add libxml files as unknown license reference test

Also fix other test failures.

Reference: https://github.com/nexB/scancode-toolkit/issues/3112
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

bwjohnson-ss commented on docs/source/explanations/license-detection-reference.rst:
```diff
@@ -0,0 +1,194 @@
+LicenseMatch data currently is based on a ``license key`` instead of being based
+on an ``license-expression``.
+
+So if there is a ``mit and apache-2.0`` license expression detected from a single
```
Is case important with license expression operands? Does `AND` usually get capitalized?

AyanSinhaMahapatra replied:
Usually no, we use upper case actually in our output license expressions but https://github.com/nexB/license-expression/ can parse these expression irrespective of upper/lower case.

AyanSinhaMahapatra committed:
Tag license intro rules correctly

Tags rules which are license intro with `is_license_intro`
as True to take advantage of the unknown license intro before
detection rule effectively.
Also updates some referenced filenames in rules.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update false positives and unknown intro heuristics

* Update false positive detection heuristics
* Update unknown intro detection heuristics
* Update test expectations
* Add ffmpeg License.md as a test for complex licenses #3113

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add unknown license reference to package dereferencing #2965 #1379

There are unknown license statements like "This file is the same
license as the package django" which refers to a package which this
file is a part of. This is fixed by extending the dereferencing logic
to look for packages that the file belongs to and using the detected
licenses from the package.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Improve unknown reference to package dereferencing #2965 #1379

* in case of unknown references being present without top-level
  detected, dereference using license detections in legalese/readme
  files at codebase root.
* add example cases from samba/samba, sugarlabs/physics, debian fusiondirectory,
  and paddlenlp as tests.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into add-license-detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into add-license-detection

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Rename `detection_rules` to `detection_log`

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update license clues heuristics

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add `rule_url` and update scancode URLs

* add a new attribute `rule_url` to LicenseMatch results.
* deleted License attributes 'scancode_data_url' and 'scancode_text_url' and
  replaced them with a single License attribute 'scancode_url'
* regen test expectations accordingly

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Restore the --unknown-licenses experimental CLI option

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Adjust unknown licenses heuristics

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Replace `package` in referenced_filename

replaces the `package` in referenced_filename with a temporary
`INHERIT_LICENSE_FROM_PACKAGE` which will be removed later in
favour of a flag.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update docstrings and improve readability

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne this is all ready!

AyanSinhaMahapatra committed:
Update changelog and docs

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on CHANGELOG.rst:
```diff
@@ -2,9 +2,67 @@ Changelog
+
+  - We now return ``license_detections`` information at the manifest file-level
+  rather than ``licenses``. This three data attributes: ``license_expression``,
+  ``detection_log`` and ``matches``. Here ``matches`` is similar to previous
```
```suggestion
  ``detection_log`` and ``matches``. Here the ``matches`` attribute is similar to previous
```

pombredanne commented on CHANGELOG.rst:
```diff
@@ -2,9 +2,67 @@ Changelog
+
+  - A new attribute ``license_clues`` was added, which has license matches same
+    as the ``matches`` field in ``license_detections``. This has license matches
+    which are mere clues and not proper detections.
```
```suggestion
    that are mere license detection clues and not proper, conclusive license detections.
```

pombredanne committed:
Improve CHANGELOG

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

AyanSinhaMahapatra committed:
Update CHANGELOG to remove duplications

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne committed:
Bump output format to v3

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update CHANGELOG

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Bump version

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>
